contracts/command-batcher/src/encoding/evm.rs

```diff
@@ -0,0 +1,477 @@
+use std::{collections::HashMap, str::FromStr};
+
+use axelar_wasm_std::Snapshot;
+use connection_router::msg::Message;
+use cosmwasm_std::{HexBinary, Uint256};
+use ethabi::{ethereum_types, Token};
+use k256::{elliptic_curve::sec1::ToEncodedPoint, PublicKey};
+use multisig::types::Signature;
+use sha3::{Digest, Keccak256};
+
+use crate::{
+    error::ContractError,
+    types::{Command, CommandBatch, CommandType, Data, Operator, Proof},
+};
+
+impl TryFrom<Message> for Command {
+    type Error = ContractError;
+
+    fn try_from(msg: Message) -> Result<Self, Self::Error> {
+        Ok(Command {
+            command_type: CommandType::ApproveContractCall, // TODO: this would change when other command types are supported
+            command_params: command_params(
+                msg.source_chain,
+                msg.source_address,
+                ethereum_types::Address::from_str(&msg.destination_address).map_err(|_| {
+                    ContractError::InvalidMessage {
+                        reason: "destination_address is not a valid EVM address".into(),
+                    }
+                })?,
+                msg.payload_hash.as_slice().try_into().map_err(|_| {
+                    ContractError::InvalidMessage {
+                        reason: "payload_hash is not a valid keccak256 hash".into(),
+                    }
+                })?,
+            ),
+            id: command_id(msg.id),
+        })
+    }
+}
+
+impl CommandBatch {
+    pub fn new(
+        messages: Vec<Message>,
+        destination_chain_id: Uint256,
+    ) -> Result<Self, ContractError> {
+        let message_ids: Vec<String> = messages.iter().map(|msg| msg.id.clone()).collect();
+
+        let data = Data {
+            destination_chain_id,
+            commands: messages
+                .into_iter()
+                .map(|msg| msg.try_into())
+                .collect::<Result<Vec<Command>, ContractError>>()?,
+        };
+        let encoded_data = data.encode();
+
+        let id = batch_id(&message_ids);
+        let msg_to_sign = msg_to_sign(&encoded_data);
+
+        Ok(Self {
+            id,
+            message_ids,
+            data,
+            msg_to_sign,
```

Comment:
Do we really need to store this as part of the `CommandBatch`? Can't we just add another method to it `signing_payload(&self)` to calculate it whenever needed?